chips/lowrisc/src/hmac.rs

```diff
@@ -88,67 +87,64 @@ impl Hmac<'_> {
             registers: base,
             client: OptionalCell::empty(),
             data: Cell::new(None),
-            data_len: Cell::new(0),
-            data_index: Cell::new(0),
             verify: Cell::new(false),
             digest: Cell::new(None),
+            cancelled: Cell::new(false),
+            busy: Cell::new(false),
         }
     }
 
-    fn data_progress(&self) -> bool {
+    fn process(&self, data: &dyn Index<usize, Output = u8>, count: usize) -> usize {
         let regs = self.registers;
-        let idx = self.data_index.get();
-        let len = self.data_len.get();
-
-        self.data.take().map_or(false, |buf| {
-            let slice = buf.take();
-
-            if idx < len {
-                let data_len = len - idx;
 
-                for i in 0..(data_len / 4) {
-                    if regs.status.is_set(STATUS::FIFO_FULL) {
-                        self.data.set(Some(LeasableBuffer::new(slice)));
-                        return false;
-                    }
+        for i in 0..(count / 4) {
+            if regs.status.is_set(STATUS::FIFO_FULL) {
+                return i * 4;
+            }
 
-                    let data_idx = idx + i * 4;
+            let data_idx = i * 4;
 
-                    let mut d = (slice[data_idx + 3] as u32) << 0;
-                    d |= (slice[data_idx + 2] as u32) << 8;
-                    d |= (slice[data_idx + 1] as u32) << 16;
-                    d |= (slice[data_idx + 0] as u32) << 24;
+            let mut d = (data[data_idx + 3] as u32) << 0;
+            d |= (data[data_idx + 2] as u32) << 8;
+            d |= (data[data_idx + 1] as u32) << 16;
+            d |= (data[data_idx + 0] as u32) << 24;
 
-                    regs.msg_fifo.set(d);
-                    self.data_index.set(data_idx + 4);
-                }
-
-                if (data_len % 4) != 0 {
-                    let idx = self.data_index.get();
+            regs.msg_fifo.set(d);
+        }
 
-                    for i in 0..(data_len % 4) {
-                        let data_idx = idx + i;
+        if (count % 4) != 0 {
+            for i in 0..(count % 4) {
+                let data_idx = (count - (count % 4)) + i;
+                regs.msg_fifo_8.set(data[data_idx]);
+            }
+        }
+        count
+    }
 
-                        regs.msg_fifo_8.set(slice[data_idx]);
-                        self.data_index.set(data_idx + 1)
-                    }
-                }
+    fn data_progress(&self) -> bool {
+        self.data.take().map_or(false, |buf| match buf {
+            LeasableBufferDynamic::Immutable(mut b) => {
+                let count = self.process(&b, b.len());
+                b.slice(count..);
+                true
+            }
+            LeasableBufferDynamic::Mutable(mut b) => {
+                let count = self.process(&b, b.len());
+                b.slice(count..);
+                true
```

Comment:
I'm a little confused. It looks like data_progress() will return `true` if we have any `self.data`. No matter if we wrote all the data or not.

Which means in `add_data()` we are going to trigger an interrupt with `INTR_TEST::FIFO_EMPTY::SET`, even if the FIFO isn't empty. Then the `self.data_progress()` in the interrupt handler will return true as well, even if we didn't write all the data.

Did you test this with a large buffer that fills the FIFO at least twice?

Reply:
Good catch. I have only been running the CI tests, as I do not have an earlgrey-cw310. The QEMU tests don't do anything.

I think this code needs to test if the length is 0 (we're done). If so, then it should return false and the caller should signal the callback. Otherwise, it should feed in the next block.

Reply:
Please take a look now.

Reply:
You can run the tests on Verilator as well, that way you can test the full system without an FPGA. It's a little slow, but pretty usable still: https://github.com/tock/tock/tree/master/boards/opentitan#verilator

Reply:
This instructions don't seem to work

https://github.com/tock/tock/tree/master/boards/opentitan#build-boot-romotp-image

Says to run ninja, but there is no build.ninja file. There doesn't seem to be an `otp_image` anymore?

Are the Tock instructions out of sync with the OpenTitan instructions?

Reply:
Did you use the correct commit?

https://github.com/tock/tock/pull/3056 has more up to date instructions, but there are still PMP and other issues to I wouldn't recommend using that

Reply:
I used the most recent bitstream. :) I then used the recommended commit and that seems to be running into trouble on the OpenTitan side, I will dig some more tonight.

Reply:
I can't find a set of instructions that works. If I go back to the recommended commit on OT  -- lowRISC/opentitan@199d45626f8a7ae2aef5d9ff73793bf9a4233711 -- there doesn't seem to be  a way to build boot_rom, which is needed in this step:

https://github.com/tock/tock/tree/master/boards/opentitan#build-and-run-tock

```shell
pal@barong:~/src/opentitan$ git status
HEAD detached at 199d45626
nothing to commit, working tree clean
pal@barong:~/src/opentitan$ grep -R boot_rom
hw/vendor/pulp_riscv_dbg/tb/boot_rom.sv:module boot_rom (
hw/vendor/pulp_riscv_dbg/tb/mm_ram.sv:    boot_rom boot_rom_i (
hw/vendor/pulp_riscv_dbg/tb/Makefile:RTLSRC_TB		:= boot_rom.sv \
hw/vendor/pulp_riscv_dbg/tb/Makefile:RTLSRC_VERI_TB          := boot_rom.sv \
Binary file .git/objects/pack/pack-0db5ff8687ec6674c9c9988e048256cba5150471.pack matches
Binary file .git/index matches
doc/ug/quickstart.md:$ ./hw/top_earlgrey/Vchip_earlgrey_verilator --rominit=./sw/device/sim/boot_rom/boot_rom_sim_verilator.scr.39.vmem --flashinit=./sw/device/sim/examples/hello_world/sw.vmem
doc/ug/quickstart.md:Rom initialized with program at ./sw/device/sim/boot_rom/rom.vmem
```

Reply:
Hmmm.... What error do you get when trying to build it with Ninja?

Reply:
Where in the instructions is the command to build it with Ninja?

Reply:
https://docs.opentitan.org/doc/getting_started/setup_fpga/

You should just be able to run

```sh
./meson_init.sh
ninja -C build-out sw/device/lib/testing/test_rom/test_rom_export_sim_verilator
ninja -C build-out sw/device/otp_img/otp_img_sim_verilator.vmem
```

to build everything you need